.github/workflows/deploy.yml

```diff
@@ -20,7 +20,29 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v5
+      
+      - name: Check deployment permissions
+        id: deploy-check
+        uses: actions/github-script@v7
+        with:
+          script: |
+            const script = require('./.github/scripts/check-deploy-permissions.js');
+            await script({ core, context });
+      
       - uses: ./.github/workflows/setup
+      
+      # Build and test steps (always run for validation)
+      - name: Build API (validation)
+        if: steps.deploy-check.outputs.should-deploy == 'false'
+        run: pnpm -C api run build
+        
+      - name: Build CDN (validation)
+        if: steps.deploy-check.outputs.should-deploy == 'false'
+        run: pnpm -C cdn run build
+        
+      - name: Build Frontend (validation)
+        if: steps.deploy-check.outputs.should-deploy == 'false'
+        run: pnpm -C frontend run build
```

Comment:
Maybe instead of adding these, we can add a flag to `./.github/workflows/deploy-worker` whether or not it should try to deploy.